### PR TITLE
Issue_1776_2: [Unit Tests] Refactor PipelineRunManagerTest. Part 2.

### PR DESCRIPTION
**Pull request by Sergeev-Ilya**
Issue: https://github.com/epam/cloud-pipeline/issues/1776

### CODE REVIEW

File: api/src/test/java/com/epam/pipeline/manager/pipeline/PipelineRunManagerUnitTest.java [hunk 1]
```diff
@@ -180,6 +204,104 @@ public void shouldLoadRunsWithTools() {
...
+        final String paramValue = String.format(paramValuePattern, ENV_VAR_NAME);
+        final String expectedValue = String.format(expectedValuePattern, ENV_VAR_VALUE);
+        final List<PipelineRunParameter> actualParameters = pipelineRunManager.replaceParametersWithEnvVars(
+                singletonList(new PipelineRunParameter(PARAM_NAME_1, paramValue)), envVars);
+
+        assertEquals(expectedValue, actualParameters.get(0).getResolvedValue());
+        assertEquals(paramValue, actualParameters.get(0).getValue());
+    }
+
+    private List<PipelineRunParameter> assertEnvVarsReplacement(final List<PipelineRunParameter> initialParameters,
```
Comment (tcibinan) [inline api/src/test/java/com/epam/pipeline/manager/pipeline/PipelineRunManagerUnitTest.java]: Usually `assertSomething` methods returns void and assert certain values. Current method doesn't assert anything and it doens't really do anything but wrap the call to `pipelineRunManager.replaceParametersWithEnvVars`. I suggest inlining the method.

File: api/src/test/java/com/epam/pipeline/manager/pipeline/PipelineRunManagerLaunchTest.java [hunk 2]
```diff
@@ -174,13 +200,20 @@ public void launchPipelineShouldValidatePriceType() {
...
     }
 
+    @Test
+    public void launchPipelineShouldValidateToolInstanceTypeAndPriceType() {
+        launchTool(configuration, INSTANCE_TYPE);
+
+        verify(instanceOfferManager).isToolInstanceAllowed(eq(INSTANCE_TYPE), eq(getResource()), eq(ID), eq(true));
+        verify(instanceOfferManager).isPriceTypeAllowed(eq(SPOT), eq(getResource()), eq(false));
+    }
+
```
Comment (tcibinan) [inline api/src/test/java/com/epam/pipeline/manager/pipeline/PipelineRunManagerLaunchTest.java]: We removed this test during the previous pull request [review](https://github.com/epam/cloud-pipeline/pull/1793#discussion_r575749367). I believe it's a problem that appeared during the conflict solving. Could you please remove this test and some other similar issues such as `getResource` private method.

File: api/src/test/java/com/epam/pipeline/manager/pipeline/PipelineRunManagerLaunchTest.java [hunk 3]
```diff
@@ -328,6 +356,28 @@ public void workerRunShouldUseParentRunCloudRegionWithParentRunIdPassedAsParamet
...
+    public void shouldLoadRunsActivityStats() {
+        final PipelineRun pipelineRun = getPipelineRun(ID, TEST_USER);
+        final PipelineRun pipelineRun2 = getPipelineRun(ID_2, TEST_USER);
+        doReturn(pipelineRun).when(pipelineRunDao).loadPipelineRun(anyLong());
+        doReturn(newArrayList(pipelineRun, pipelineRun2)).when(pipelineRunDao)
+                .loadPipelineRunsActiveInPeriod(eq(TEST_PERIOD), eq(TEST_PERIOD.plusHours(HOURS_18)));
+        doReturn(getStatusMap()).when(runStatusManager).loadRunStatus(anyListOf(Long.class));
+
+        launchPipelineRun(TEST_PERIOD, TEST_PERIOD.plusHours(HOURS_12));
+        launchPipelineRun(TEST_PERIOD.plusHours(HOURS_18), null);
```
Comment (tcibinan) [inline api/src/test/java/com/epam/pipeline/manager/pipeline/PipelineRunManagerLaunchTest.java]: As long as we write a unit test that tests a single `pipelineRunManager.loadRunsActivityStats` method please mock and verify execution of a single method only. No need to launch any pipelines within fully mocked environment. As a result the test should be significantly simplified.

File: api/src/test/java/com/epam/pipeline/manager/pipeline/PipelineRunManagerLaunchTest.java [hunk 4]
```diff
@@ -368,6 +418,36 @@ private void mock(final Tool tool) {
...
+    private Map<Long, List<RunStatus>> getStatusMap() {
+        final RunStatus status1 = new RunStatus(ID, TaskStatus.RUNNING, null, TEST_PERIOD);
+        final RunStatus status2 = new RunStatus(ID, TaskStatus.STOPPED, null, TEST_PERIOD.plusHours(HOURS_18));
+        final List<RunStatus> list1 = newArrayList(status1, status2);
+        final RunStatus status3 = new RunStatus(ID_2, TaskStatus.RUNNING, null, TEST_PERIOD.plusHours(HOURS_18));
+        final List<RunStatus> list2 = newArrayList(status3);
+        final Map<Long, List<RunStatus>> map = new HashMap<>();
+        map.put(ID, list1);
+        map.put(ID_2, list2);
+        return map;
```
Comment (tcibinan) [inline api/src/test/java/com/epam/pipeline/manager/pipeline/PipelineRunManagerLaunchTest.java]: I believe we can greatly simplify such structures avoiding local environment variables:

```java
private Map<Long, List<RunStatus>> getStatusMap() {
    final Map<Long, List<RunStatus>> map = new HashMap<>();
    map.put(ID, Arrays.asList(
            new RunStatus(ID, TaskStatus.RUNNING, null, TEST_PERIOD),
            new RunStatus(ID, TaskStatus.STOPPED, null, TEST_PERIOD.plusHours(HOURS_18))));
    map.put(ID_2, Collections.singletonList(
            new RunStatus(ID_2, TaskStatus.RUNNING, null, TEST_PERIOD.plusHours(HOURS_18))));
    return map;
}
```

File: api/src/test/java/com/epam/pipeline/manager/pipeline/PipelineRunManagerLaunchTest.java [hunk 5]
```diff
@@ -368,6 +418,36 @@ private void mock(final Tool tool) {
...
+        } else {
+            pipelineRun.setEndDate(null);
+        }
+        pipelineRunManager.updateRunInfo(pipelineRun);
+    }
+
+    private Map<Long, List<RunStatus>> getStatusMap() {
+        final RunStatus status1 = new RunStatus(ID, TaskStatus.RUNNING, null, TEST_PERIOD);
+        final RunStatus status2 = new RunStatus(ID, TaskStatus.STOPPED, null, TEST_PERIOD.plusHours(HOURS_18));
+        final List<RunStatus> list1 = newArrayList(status1, status2);
```
Comment (tcibinan) [inline api/src/test/java/com/epam/pipeline/manager/pipeline/PipelineRunManagerLaunchTest.java]: Also please take attention to which static methods you use. Currently you have an import for `org.mapstruct.ap.internal.util.Collections.newArrayList` which is just some internal static method from some random library in classpath. Please use canonical `java.util` static methods if possible.

**Review by tcibinan — CHANGES_REQUESTED** [review]

File: api/src/test/java/com/epam/pipeline/manager/pipeline/PipelineRunManagerLaunchTest.java (hunk 5 above)
Comment (Sergeev-Ilya) [reply api/src/test/java/com/epam/pipeline/manager/pipeline/PipelineRunManagerLaunchTest.java]: Done!

File: api/src/test/java/com/epam/pipeline/manager/pipeline/PipelineRunManagerLaunchTest.java (hunk 4 above)
Comment (Sergeev-Ilya) [reply api/src/test/java/com/epam/pipeline/manager/pipeline/PipelineRunManagerLaunchTest.java]: Done!

File: api/src/test/java/com/epam/pipeline/manager/pipeline/PipelineRunManagerLaunchTest.java (hunk 3 above)
Comment (Sergeev-Ilya) [reply api/src/test/java/com/epam/pipeline/manager/pipeline/PipelineRunManagerLaunchTest.java]: Done!

File: api/src/test/java/com/epam/pipeline/manager/pipeline/PipelineRunManagerLaunchTest.java (hunk 2 above)
Comment (Sergeev-Ilya) [reply api/src/test/java/com/epam/pipeline/manager/pipeline/PipelineRunManagerLaunchTest.java]: Done!

File: api/src/test/java/com/epam/pipeline/manager/pipeline/PipelineRunManagerUnitTest.java (hunk 1 above)
Comment (Sergeev-Ilya) [reply api/src/test/java/com/epam/pipeline/manager/pipeline/PipelineRunManagerUnitTest.java]: Done!

File: api/src/test/java/com/epam/pipeline/manager/pipeline/PipelineRunManagerLaunchTest.java (hunk 5 above)
Comment (Sergeev-Ilya) [reply api/src/test/java/com/epam/pipeline/manager/pipeline/PipelineRunManagerLaunchTest.java]: Done!

**Comment by Sergeev-Ilya** [comment]
Solved all the comments

File: api/src/test/java/com/epam/pipeline/manager/pipeline/PipelineRunManagerLaunchTest.java [hunk 6]
```diff
@@ -328,6 +347,19 @@ public void workerRunShouldUseParentRunCloudRegionWithParentRunIdPassedAsParamet
...
+        doReturn(Arrays.asList(getPipelineRun(ID, TEST_USER), getPipelineRun(ID_2, TEST_USER)))
+                .when(pipelineRunDao).loadPipelineRunsActiveInPeriod(eq(TEST_PERIOD), eq(TEST_PERIOD_18));
+        doReturn(getStatusMap()).when(runStatusManager).loadRunStatus(anyListOf(Long.class));
+
+        pipelineRunManager.loadRunsActivityStats(TEST_PERIOD, TEST_PERIOD_18).stream()
+                .collect(Collectors.toMap(BaseEntity::getId, Function.identity()));
+
+        verify(pipelineRunDao).loadPipelineRunsActiveInPeriod(any(LocalDateTime.class), any(LocalDateTime.class));
+        verify(runStatusManager).loadRunStatus(anyListOf(Long.class));
+    }
```
Comment (tcibinan) [inline api/src/test/java/com/epam/pipeline/manager/pipeline/PipelineRunManagerLaunchTest.java]: If you look at this test after all the changes applied it doesn't look like it tests anything what `pipelineRunManager.loadRunsActivityStats` really does. At this point there is no sense using `getStatusMap()` you could replace it with `Collections.emptyMap()` and the test will still pass.

The whole idea behind `pipelineRunManager.loadRunsActivityStats` is the following:
1. It loads active runs.
2. It loads statuses for found active runs.
3. It merges loaded runs with loaded statuses.

As a result in the test you should mock 1 and 2 points and verify the 3. You should verify that the runs returned by the target method have associated run statuses. You could even skip `verify` calls if you check the returned runs instead. That was the idea behind the original test and we don't have to change it.

The test should look something like
```java
mockRuns
mockStatuses

loadRunsActivityStats

assertLoadedRunsHaveAssociatedStatuses
```

**Review by tcibinan — CHANGES_REQUESTED** [review]

File: api/src/test/java/com/epam/pipeline/manager/pipeline/PipelineRunManagerLaunchTest.java (hunk 6 above)
Comment (Sergeev-Ilya) [reply api/src/test/java/com/epam/pipeline/manager/pipeline/PipelineRunManagerLaunchTest.java]: Resolved!

File: api/src/test/java/com/epam/pipeline/manager/pipeline/PipelineRunManagerLaunchTest.java [hunk 7]
```diff
@@ -368,6 +409,13 @@ private void mock(final Tool tool) {
         doReturn(tool).when(toolManager).resolveSymlinks(eq(IMAGE));
     }
 
+    private Map<Long, List<RunStatus>> getStatusMap() {
+        final Map<Long, List<RunStatus>> map = new HashMap<>();
+        map.put(ID, Arrays.asList(TEST_STATUS_1, TEST_STATUS_2));
+        map.put(ID_2, singletonList(TEST_STATUS_3));
```
Comment (tcibinan) [inline api/src/test/java/com/epam/pipeline/manager/pipeline/PipelineRunManagerLaunchTest.java]: It feels odd to have static import for `Collections.singletonList` and to don't have it for `Arrays.asList`. I would recommend using either one syntax or another not both.

```java
# bad
map.put(ID, Arrays.asList(TEST_STATUS_1, TEST_STATUS_2));
map.put(ID_2, singletonList(TEST_STATUS_3));

# good
map.put(ID, Arrays.asList(TEST_STATUS_1, TEST_STATUS_2));
map.put(ID_2, Collections.singletonList(TEST_STATUS_3));

# good too
map.put(ID, asList(TEST_STATUS_1, TEST_STATUS_2));
map.put(ID_2, singletonList(TEST_STATUS_3));
```

**Review by tcibinan — CHANGES_REQUESTED** [review]

File: api/src/test/java/com/epam/pipeline/manager/pipeline/PipelineRunManagerLaunchTest.java (hunk 7 above)
Comment (Sergeev-Ilya) [reply api/src/test/java/com/epam/pipeline/manager/pipeline/PipelineRunManagerLaunchTest.java]: Done!

**Review by tcibinan — APPROVED** [review]

**Pull request merged**